### ISSUE TITLE
Preserve the background information in training process.

### ISSUE THREAD

**Issue by MissDores** [eladrich/pixel2style2pixel#109]
Thank you very much for the great work and the release of code!

There are three questions I want to ask for your advices:
1. With the identity loss, the face preserve the detials from input perfectly.  I wonder know if there are some tricks and mothods to keep the background information as well.
2. The pretrained stylegan decoder can generate a face much more real, but most of them are not occluded and frontalizated. If I want to generate real samples which are occluded or with big posture  , do I need to train a new stylegan decoder,   or just according to adjust the loss for more fidelity,  or perhaps with just paired such samples without any other changes?
3. If I need to train a new stylegan decoder,  should I need to use the stylegan2 project or just keep the train_decoder flag to True? As I do not see the use of stylegan discriminator and related gan loss in psp project.

Thank you very much! @yuval-alaluf

**Comment by yuval-alaluf** [comment]
Hi @MissDores ,
1. Regarding the preservation of background information, this is a bit trickier than preserving facial features because sometimes this entails preserving fine details such as text which is challenging to do when encoding into style vectors. You could try incorporating a patch-based loss and compare random patches from the input and reconstructed image (both in the pixel space and activation space). I haven't tried this myself, but this could be interesting. 
2. If the generator you're using does not know how to generate facial images with many occlusions or difficult poses, you may need to fine-tune it on more challenging data. Adding an additional loss or adjusting the current losses won't necessarily help if the generator cannot generate such images. 
3.  If you need to train/fine-tune your generator, I would recommend using the stylegan2 project either from rosinality or NVIDIA (which you will need to convert to PyTorch afterward). 
I hope this addresses your questions.

**Comment by MissDores** [comment]
Thanks for your kind reply, it's very helpful!

**Comment by MissDores** [comment]
Thanks for your kind reply, it's very helpful!

**Comment by MissDores** [comment]
Thanks for your kind reply, it's very helpful!
One more confusion is that, in which situation do i need to set train_decoder flag "True" and this will help?
And as for the loss tuning in this project setting,   can you offer some guidance for super resolution of the tuning effects of every parameters?

**Comment by yuval-alaluf** [comment]
We didn't find a scenario where we needed to train the decoder with the encoder, but it could be interesting if your data is quite different than the data the original generator was trained on.  
Regarding the losses of super resolution, these are the same weights used in all other applications. Please look at the paper where we tried to provide some insights to the loss functions. 
You can take a look at https://github.com/eladrich/pixel2style2pixel/issues/49 for details regarding the w-norm loss.

**Comment by MissDores** [comment]
Sincere thanks!
